### PR TITLE
Create containerized CI testing

### PR DESCRIPTION
**Pull request by ggeorgakoudis**
Create a workflow to run tests inside a container that mimics the ruby environment.

### CODE REVIEW

File: .github/containers/ruby/repo/use-mirror/mirrors.yaml
Comment (koparasy) [inline .github/containers/ruby/repo/use-mirror/mirrors.yaml:1]: Do you need this file? Are we pulling from the root mirror?

File: .github/containers/ruby/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# configuration settings.
+spack:
+  # add package specs to the `specs` list
+  specs:
+  - python
+  - py-torch
+  - py-scipy
+  - py-torchvision
+  - py-horovod
+  - py-pillow
```
Comment (koparasy) [inline .github/containers/ruby/spack.yaml]: Remove. AMSLib does not use those AFAIK.

File: .github/containers/ruby/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  - py-h5py
+  - py-pandas
+  - py-scikit-learn
+  - py-matplotlib
```
Comment (koparasy) [inline .github/containers/ruby/spack.yaml]: Remove

File: .github/containers/ruby/repo/packages/py-hdbscan/package.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from spack.package import *
+
+
+class PyHdbscan(PythonPackage):
```
Comment (koparasy) [inline .github/containers/ruby/repo/packages/py-hdbscan/package.py]: Why do we need this package? Who is using it?

**Comment by ggeorgakoudis** [comment]
I've copied spack.yaml and repo structure verbatim from the spack setup on the shared workspace. Happy to remove them, rebuild the container and test, but it means this setup will diverge from the shared workspace. Are you OK with that?

File: .github/containers/ruby/repo/packages/py-llnl-sina/package.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
```
Comment (koparasy) [inline .github/containers/ruby/repo/packages/py-llnl-sina/package.py]: I am not sure we can have these lines in our repository. These refers to the spack license and not ours.

File: .github/containers/ruby/repo/packages/flux-sched/package.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
```
Comment (koparasy) [inline .github/containers/ruby/repo/packages/flux-sched/package.py]: I am not sure we can have the spack license in our repo. This applies to all packages you are including.

**Review by koparasy — CHANGES_REQUESTED** [review]
Thank you. This overall is good. We need some minor restructuring to make this work for other "containers". 

This seems as a "Generic" configuration to test AMS using a "RUBY" specific spack/container structure. However, most of the spack related package patches are way more general than that. I suggest you move the "repo" directory to the parent (.github/containers/repo). I would also rename the "ruby" directory to x86_gcc_11.2.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+          cd build
+          export AMS_MFEM_PATH=`spack location -i mfem`
+          export AMS_TORCH_PATH=$(spack location -i py-torch)/lib/python3.10/site-packages/torch/share/cmake/Torch
+          export AMS_FAISS_PATH=`spack location -i faiss`
+          export AMS_UMPIRE_PATH=`spack location -i umpire`
+          export AMS_HIREDIS_PATH=`spack location -i hiredis`
+          export AMS_REDIS_PLUS_PLUS_PATH=`spack location -i redis-plus-plus`
+          export AMS_HDF5_PATH=`spack location -i hdf5`
+          cmake \
+            -Damqpcpp_DIR=$AMS_AMQPCPP_PATH \
```
Comment (koparasy) [inline .github/workflows/ci.yml]: Remove, we are not enabling RMQ DB.

File: tests/CMakeLists.txt [hunk 7]
```diff
@@ -23,4 +23,3 @@ endfunction()
 ADDTEST(ams_allocator ams_allocate.cpp AMSAllocate)
 ADDTEST(ams_packing cpu_packing_test.cpp AMSPack)
 ADDTEST(ams_inference torch_model.cpp AMSInfer /usr/workspace/AMS/miniapp_resources/trained_models/debug_model.pt)
```
Comment (koparasy) [inline tests/CMakeLists.txt]: Absolute path should be fixed

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+          module load gcc/11.2.1
+          source /spack/share/spack/setup-env.sh
+          spack env activate -p /ams-spack-env
+          mkdir build
+          cd build
+          export AMS_MFEM_PATH=`spack location -i mfem`
+          export AMS_TORCH_PATH=$(spack location -i py-torch)/lib/python3.10/site-packages/torch/share/cmake/Torch
+          export AMS_FAISS_PATH=`spack location -i faiss`
+          export AMS_UMPIRE_PATH=`spack location -i umpire`
+          export AMS_HIREDIS_PATH=`spack location -i hiredis`
```
Comment (ggeorgakoudis) [inline .github/workflows/ci.yml]: @koparasy This is redundant too, remove?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+          export AMS_REDIS_PLUS_PLUS_PATH=`spack location -i redis-plus-plus`
```
Comment (ggeorgakoudis) [inline .github/workflows/ci.yml]: @koparasy This is redundant too, remove?

File: .github/containers/ruby/spack.yaml [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  - py-pip
+  - py-kosh
+  - py-mpi4py
+  - mfem
+  - umpire
+  - caliper
+  - faiss
+  - redis-plus-plus
+  - amqp-cpp +tcp
```
Comment (ggeorgakoudis) [inline .github/containers/ruby/spack.yaml]: @koparasy Those can be removed too?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (koparasy) [reply .github/workflows/ci.yml]: No we decided to have it there as it is closer to what we actually do.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (koparasy) [reply .github/workflows/ci.yml]: Same as above.

File: .github/containers/ruby/spack.yaml (hunk 10 above)
Comment (koparasy) [reply .github/containers/ruby/spack.yaml]: redis++ for now yes, amqp needs to be there when we enable RMQ via build options.

File: .github/containers/ruby/repo/use-mirror/mirrors.yaml
Comment (ggeorgakoudis) [reply .github/containers/ruby/repo/use-mirror/mirrors.yaml:1]: removed

File: .github/containers/ruby/spack.yaml (hunk 1 above)
Comment (ggeorgakoudis) [reply .github/containers/ruby/spack.yaml]: removed

File: .github/containers/ruby/repo/packages/py-hdbscan/package.py (hunk 3 above)
Comment (ggeorgakoudis) [reply .github/containers/ruby/repo/packages/py-hdbscan/package.py]: needed by dependency

File: .github/containers/ruby/repo/packages/flux-sched/package.py (hunk 5 above)
Comment (ggeorgakoudis) [reply .github/containers/ruby/repo/packages/flux-sched/package.py]: changed to AMS

File: .github/workflows/ci.yml (hunk 6 above)
Comment (ggeorgakoudis) [reply .github/workflows/ci.yml]: removed

File: .github/containers/ruby/spack.yaml (hunk 2 above)
Comment (ggeorgakoudis) [reply .github/containers/ruby/spack.yaml]: removed

File: .github/containers/ruby/repo/packages/py-llnl-sina/package.py (hunk 4 above)
Comment (ggeorgakoudis) [reply .github/containers/ruby/repo/packages/py-llnl-sina/package.py]: changed to AMS license

File: .github/workflows/ci.yml (hunk 9 above)
Comment (ggeorgakoudis) [reply .github/workflows/ci.yml]: superseded by above comment

File: .github/workflows/ci.yml (hunk 8 above)
Comment (ggeorgakoudis) [reply .github/workflows/ci.yml]: The comment refes to `AMS_HIREDDIS_PATH`. The package hiredis is not installed so it's buggy to leave it there. We can always add it in a later version

Comment (koparasy) [reply .github/workflows/ci.yml]: Yes.`AMS_HIREDDIS_PATH` should be removed.

**Review by koparasy — APPROVED** [review]

**Pull request merged**